src/magicbox/ResultPreviewsManager.ts

```diff
@@ -0,0 +1,214 @@
+import { $$, Dom } from '../utils/Dom';
+import { l } from '../strings/Strings';
+import { defaults, findIndex } from 'lodash';
+import { Component } from '../ui/Base/Component';
+import { Direction } from './SuggestionsManager';
+
+export interface IPopulateSearchResultPreviewsEventArgs {
+  suggestionText: string;
+  previewsQuery: Promise<ISearchResultPreview[]>;
+}
+
+export enum ResultPreviewsManagerEvents {
+  PopulateSearchResultPreviews = 'populateSearchResultPreviews'
+}
+
+export interface ISearchResultPreview {
+  element: HTMLElement;
+  onSelect: () => void;
+}
+
+export interface IResultPreviewsManagerOptions {
+  previewClass: string;
+  selectedClass: string;
+  previewHeaderText: string;
+}
+
+export class ResultPreviewsManager {
+  private options: IResultPreviewsManagerOptions;
+  private suggestionsPreviewContainer: Dom;
+  private resultPreviewsHeader: Dom;
+  private resultPreviewsContainer: Dom;
+  private lastPreviewsQuery: Promise<ISearchResultPreview[]>;
+  private lastSelectedSuggestion: HTMLElement;
+  private root: HTMLElement;
+
+  public get previewsOwner() {
+    return this.lastSelectedSuggestion;
+  }
+
+  public get hasPreviews() {
+    return !!this.suggestionsPreviewContainer;
+  }
+
+  public get focusedPreviewElement() {
+    if (!this.hasPreviews) {
+      return null;
+    }
+    const focusedElement = this.suggestionsPreviewContainer.findClass(this.options.selectedClass)[0];
+    if (!focusedElement || !focusedElement.classList.contains(this.options.previewClass)) {
+      return null;
+    }
+    return focusedElement;
+  }
+
+  public get previewElements() {
+    if (!this.hasPreviews) {
+      return [];
+    }
+    return this.suggestionsPreviewContainer.findClass(this.options.previewClass);
+  }
+
+  private get suggestionsListbox() {
+    return $$($$(this.element).findClass('coveo-magicbox-suggestions')[0]);
+  }
+
+  private get numberOfResultsPerRow() {
+    const previewSelectables = this.suggestionsPreviewContainer.findClass(this.options.previewClass);
+    if (previewSelectables.length === 0) {
+      return 0;
+    }
+    const firstVerticalOffset = previewSelectables[0].offsetTop;
+    const firstIndexOnNextRow = findIndex(previewSelectables, previewSelectable => previewSelectable.offsetTop !== firstVerticalOffset);
+    return firstIndexOnNextRow !== -1 ? firstIndexOnNextRow : previewSelectables.length;
+  }
+
+  constructor(private element: HTMLElement, options: Partial<IResultPreviewsManagerOptions> = {}) {
+    this.options = defaults(options, <IResultPreviewsManagerOptions>{
+      previewHeaderText: l('QuerySuggestPreview'),
+      previewClass: 'coveo-preview-selectable',
+      selectedClass: 'magic-box-selected'
+    });
```

Comment:
Unfortunately, using the spread operator instead of this would mean that values could be overwritten by `undefined`. Weird behaviour.

Reply:
You mean by doing this?
```
this.options = {
  ...options,
  previewHeaderText: l('QuerySuggestPreview'),
  previewClass: 'coveo-preview-selectable',
  selectedClass: 'magic-box-selected'
}
```

Reply:
Yes, unfortunately. I don't remember exactly the case where an `undefined` would be passed, but it caused issues. Maybe in this case it would be fine?

Reply:
I have no prevents, if you prefer to use defaults go for it!